### PR TITLE
Implement a basic keybinding collision checker. (ToDo list #337)

### PR DESCRIPTION
**Pull request by annagrram**

### CODE REVIEW

**Comment by jarun** [comment]
Wow! This was quick! Thanks for the PR. I will this review shortly.

Note for @jarun: Please document this in the Wiki Advanced...

File: misc/keybind-collision-checker/collision.c [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+} bucket_t, *pbucket_t;
+
+static bool is_collision_in_bucket(pbucket_t buckets, int sym);
+static void free_buckets(pbucket_t buckets);
+
+int main() {
+    bucket_t buckets[BUCKETS_ARRAY_SIZE] = {0};
+    bool collision_detected = FALSE;
+
+    for (int i = 0; i < sizeof(bindings) / sizeof(struct key); ++i) {
```
Comment (jarun) [inline misc/keybind-collision-checker/collision.c]: Please declare and initialize `i` at the top of the function. Compiler warns.

File: misc/keybind-collision-checker/collision.c [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        pbucket_t curr_bucket = &buckets[curr_sym % BUCKETS_ARRAY_SIZE];
+
+        if (is_collision_in_bucket(curr_bucket, curr_sym)) {
+            printf("Collision of key %s detected\n", keyname(curr_sym));
+
+            collision_detected = TRUE;
+        }
+    }
+
+    if (!collision_detected)
```
Comment (jarun) [inline misc/keybind-collision-checker/collision.c]: I think we can remove this. Notifying only if a collision is detected should be enough. No news is good news.

Comment (jarun) [reply misc/keybind-collision-checker/collision.c]: Get rid of the boolean too.

File: misc/keybind-collision-checker/collision.c [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    bucket_t buckets[BUCKETS_ARRAY_SIZE] = {0};
+    bool collision_detected = FALSE;
+
+    for (int i = 0; i < sizeof(bindings) / sizeof(struct key); ++i) {
+        int curr_sym = bindings[i].sym;
+        pbucket_t curr_bucket = &buckets[curr_sym % BUCKETS_ARRAY_SIZE];
+
+        if (is_collision_in_bucket(curr_bucket, curr_sym)) {
+            printf("Collision of key %s detected\n", keyname(curr_sym));
+
```
Comment (jarun) [inline misc/keybind-collision-checker/collision.c]: Blank line not required.

File: misc/keybind-collision-checker/collision.c [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    free_buckets(buckets);
+}
+
+bool is_collision_in_bucket(pbucket_t bucket, int sym) {
+    while (bucket) {
+        if (bucket->sym == sym) {
+            return TRUE;
+        } else if (bucket->sym == 0) {
+            bucket->sym = sym;
+
```
Comment (jarun) [inline misc/keybind-collision-checker/collision.c]: Blank line not required.

File: misc/keybind-collision-checker/collision.c [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+            break;
+        }
+        else if (!bucket->next) {
```
Comment (jarun) [inline misc/keybind-collision-checker/collision.c]: Keep the `else if` along with the bracket.

File: misc/keybind-collision-checker/collision.c [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+            bucket->next = (pbucket_t)malloc(sizeof(bucket_t));
+            bucket->next->sym = sym;
+            bucket->next->next = NULL;
+
```
Comment (jarun) [inline misc/keybind-collision-checker/collision.c]: Redundant bank line.

File: misc/keybind-collision-checker/collision.c [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        }
+    }
+
+    if (!collision_detected)
+        printf("No collisions detected\n");
+
+    free_buckets(buckets);
+}
+
+bool is_collision_in_bucket(pbucket_t bucket, int sym) {
```
Comment (jarun) [inline misc/keybind-collision-checker/collision.c]: Can we implement this as a cute hash (bitmap)? The value of the `key` should be the bit number in the bitmap. Set if not set already. If set, you have a collision.

File: misc/keybind-collision-checker/collision.c [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+        } else {
+            bucket = bucket->next;
+        }
+    }
+
+    return FALSE;
+}
+
+void free_buckets(pbucket_t buckets) {
+    for (int i = 0; i < BUCKETS_ARRAY_SIZE; ++i) {
```
Comment (jarun) [inline misc/keybind-collision-checker/collision.c]: `int i = 0` goes outside.

File: misc/keybind-collision-checker/collision.c (hunk 7 above)
Comment (jarun) [reply misc/keybind-collision-checker/collision.c]: Note: please check the feasibility. I am not sure how far the CONTROL+char combinations can scale to. If we can have a max value (up to) 10000, the bitmap would be ~1.25K. Good enough!

Comment (annagrram) [reply misc/keybind-collision-checker/collision.c]: Well, that made everything simpler :)
Changed that to a boolean array for more straightforward implementation.

File: misc/keybind-collision-checker/collision.c [hunk 9]
```diff
@@ -0,0 +1,20 @@
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "../../src/nnn.h"
+
+int main() {
+    int i;
+
+    // CONTROL(KEY_MAX) == 447
+    bool bitmap[CONTROL(KEY_MAX)] = {FALSE};
```
Comment (jarun) [inline misc/keybind-collision-checker/collision.c]: bool is unsigned char internally. Can we have a true bitmap where each key is represented by 1 bit (and not byte)?

The bitmap size would be: ((CONTROL(KEY_MAX) + 7) / 8).

File: misc/keybind-collision-checker/collision.c [hunk 10]
```diff
@@ -0,0 +1,20 @@
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "../../src/nnn.h"
```
Comment (jarun) [inline misc/keybind-collision-checker/collision.c]: When it's ready, we will have it in `nnn.c` and expose an option to test collision.

File: misc/keybind-collision-checker/collision.c (hunk 9 above)
Comment (jarun) [reply misc/keybind-collision-checker/collision.c]: Also, on my system:

`KEY_MAX`: 511
`CONTROL(KEY_MAX)` = 447

Probably we should use `KEY_MAX`.

Comment (annagrram) [reply misc/keybind-collision-checker/collision.c]: I'll check again and change the length if necessary. Thanks for noticing.

I don't think there's a simple way of doing such a big bitmap in plain C.
I guess the easiest way would be to create an array with the correct number of bits and each time to calculate in which cell the bit should be and then calculate which bit should be checked or flipped and do it.

I'm not sure this increase in code complexity is worth the relatively minor difference in memory.
Modern and less than modern system should handle an array of bytes of length 511 quite easily. Especially when the function that creates it is short-lived and the memory isn't occupied for long.

What do you think?

Comment (jarun) [reply misc/keybind-collision-checker/collision.c]: > Modern and less than modern systems should handle an array of bytes of length 511 quite easily

True. There's not a huge save there.

Can you add it in nnn.c and add a program option?

Comment (annagrram) [reply misc/keybind-collision-checker/collision.c]: Sure.
Will fix and commit soon.

Comment (annagrram) [reply misc/keybind-collision-checker/collision.c]: Done.

I chose the flag K, feel free to change it.

**Comment by jarun** [comment]
Thank you for your contribution!

**Comment by jarun** [comment]
@annagrram could you make some time to add the the patch for coreutils v8.31 to [advcmpv](https://github.com/jarun/advcpmv)?

**Comment by annagrram** [comment]
Sure.
I'll take a look later today.

**Pull request merged**